File: Rust/src/apple.rs
use crate::{rng::get_random_int, snake::Snake};

pub struct Apple {
    pub x: i32,
    pub y: i32,
}

impl Apple {
    pub fn new(snake: &Snake) -> Self {
        let mut apple = Apple { x: 0, y: 0 };
        apple.eat(snake);
        apple
    }

    pub fn eat(&mut self, snake: &Snake) {
        for _ in 0..1000 {
            let x: i32 = get_random_int(0, 19);
            let y: i32 = get_random_int(0, 19);

            let mut blocked: bool = false;
            let mut curr: Option<&crate::node::Node> = Some(&*snake.head);

            while let Some(node) = curr {
                if node.x == x && node.y == y {
                    blocked = true;
                    break;
                }
                curr = node.next.as_deref();
            }

            if !blocked {
                self.x = x;
                self.y = y;
                break;
            }
        }
    }
}

File: Typescript/src/apple.ts
import getRandomInt from "./random.js";
import type Snake from "./snake.js";

export default class Apple {
    public x: number = 0;
    public y: number = 0;

    constructor(snake: Snake) {
        this.eat(snake);
    }

    public eat(snake: Snake) {
        for (let i = 0; i < 1000; ++i) {
            let blocked = false;

            const rx = getRandomInt(0, 19);
            const ry = getRandomInt(0, 19);

            for (const node of snake.head) {
                if (node.x == rx && node.y == ry) {
                    blocked = true;
                    break;
                }
            }

            if (!blocked) {
                this.x = rx;
                this.y = ry;
                return;
            }
        }
    }
}
